tfhe/src/integer/server_key/radix_parallel/scalar_sub.rs

```diff
@@ -124,27 +126,477 @@ impl ServerKey {
             self.full_propagate_parallelized(lhs);
         }
 
-        let mut scalar_decomposer =
-            BlockDecomposer::new(scalar, self.message_modulus().0.ilog2()).iter_as::<u8>();
+        if self.is_eligible_for_parallel_single_carry_propagation(lhs.blocks.len()) {
+            self.unsigned_overflowing_scalar_sub_assign_parallelized_at_least_4_bits(lhs, scalar)
+        } else {
+            self.unsigned_overflowing_scalar_sub_assign_parallelized_sequential(lhs, scalar)
+        }
+    }
+
+    fn unsigned_overflowing_scalar_sub_assign_parallelized_sequential<Scalar>(
+        &self,
+        lhs: &mut RadixCiphertext,
+        scalar: Scalar,
+    ) -> BooleanBlock
+    where
+        Scalar: UnsignedNumeric + DecomposableInto<u8> + std::ops::Not<Output = Scalar>,
+    {
+        assert!(!lhs.blocks.is_empty(), "lhs cannot be empty");
+
+        let scalar_blocks = BlockDecomposer::new(scalar, self.message_modulus().0.ilog2())
+            .iter_as::<u8>()
+            .take(lhs.blocks.len())
+            .collect::<Vec<_>>();
+
+        // If the block does not have a carry after the subtraction, it means it needs to
+        // borrow from the next block
+        let compute_borrow_lut = self.key.generate_lookup_table(|x| {
+            if x < self.message_modulus().0 as u64 {
+                1
+            } else {
+                0
+            }
+        });
+
+        let mut borrow = self.key.create_trivial(0);
+        let delta = (1_u64 << 63) / (self.message_modulus().0 * self.carry_modulus().0) as u64;
+        for (lhs_b, scalar_b) in lhs.blocks.iter_mut().zip(scalar_blocks.iter().copied()) {
+            // Here we use core_crypto instead of shortint scalar_sub_assign
+            // because we need a true subtraction, not an addition of the inverse
+            crate::core_crypto::algorithms::lwe_ciphertext_plaintext_sub_assign(
+                &mut lhs_b.ct,
+                crate::core_crypto::prelude::Plaintext(u64::from(scalar_b) * delta),
+            );
+            crate::core_crypto::algorithms::lwe_ciphertext_plaintext_add_assign(
+                &mut lhs_b.ct,
+                crate::core_crypto::prelude::Plaintext(self.message_modulus().0 as u64 * delta),
+            );
+            lhs_b.degree = crate::shortint::ciphertext::Degree::new(
+                lhs_b.degree.get() - usize::from(scalar_b) + self.message_modulus().0,
+            );
```

Comment:
this one is likely correct (as we are subtracting scalar of which we know exactly the value) though we could have overflows during the subtraction ?